### PR TITLE
Create initial structure

### PR DESCRIPTION
**Pull request by Marenz**
pylint still makes some problems, apparently it can't find the generated files?

### CODE REVIEW

**Comment by Marenz** [comment]
@llucax is there a trick to make pylint happy or do I need to add exceptions for most of those?

```
nox > pylint src docs noxfile.py tests
************* Module frequenz.client.dispatch._client
# This one is clear
src/frequenz/client/dispatch/_client.py:100:9: W0511: TODO handle result (fixme)
src/frequenz/client/dispatch/_client.py:11:0: E0611: No name 'Timestamp' in module 'google.protobuf.timestamp_pb2' (no-name-in-module)
src/frequenz/client/dispatch/_client.py:72:28: E1101: Module 'frequenz.api.dispatch.v1.dispatch_pb2' has no 'TimeIntervalFilter' member (no-member)
src/frequenz/client/dispatch/_client.py:83:33: E1101: Module 'frequenz.api.dispatch.v1.dispatch_pb2' has no 'ComponentSelector' member (no-member)
src/frequenz/client/dispatch/_client.py:90:18: E1101: Module 'frequenz.api.dispatch.v1.dispatch_pb2' has no 'DispatchFilter' member (no-member)
src/frequenz/client/dispatch/_client.py:96:18: E1101: Module 'frequenz.api.dispatch.v1.dispatch_pb2' has no 'DispatchListRequest' member (no-member)
************* Module test_dispatch_client
# This one is clear
tests/test_dispatch_client.py:9:4: C0415: Import outside toplevel (frequenz.client.dispatch.Client) (import-outside-toplevel)
```

**Comment by llucax** [comment]
It is a known bug, we need to put an ignore in the import line:

* https://github.com/pylint-dev/pylint/issues/6281

File: pyproject.toml [hunk 1]
```diff
@@ -26,9 +26,14 @@ classifiers = [
...
-  "frequenz-api-dispatch >= 0.12.0, < 0.13",
-  "frequenz-client-base >= 0.1.0, < 0.2",
+  "typing-extensions >= 4.6.1, < 5",
+# Use marenz fork @ 0d04bc3 until the next release
+  "frequenz-api-dispatch @ git+https://github.com/marenz/frequenz-api-dispatch.git@0d04bc3",
+#  "frequenz-api-dispatch >= 0.12.0, < 0.13",
+# Directly use unreleased/unmerged commit fe4e238 until the next release
+  "frequenz-client-base @ git+https://github.com/marenz/frequenz-client-base-python.git@fe4e238",
+#  "frequenz-client-base >= 0.1.0, < 0.2",
+  "frequenz-sdk == v1.0.0-rc4",
```
Comment (llucax) [inline pyproject.toml]: We shouldn't import the SDK in API clients, the idea is to do it the other way around (the SDK will import the client).

File: src/frequenz/client/dispatch/_client.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
+# License: MIT
+# Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+"""Dispatch API client for Python."""
+from dataclasses import dataclass
+from datetime import datetime
+
+import grpc
+from frequenz.api.dispatch.v1 import dispatch_pb2, dispatch_pb2_grpc
+from frequenz.sdk.microgrid.component import ComponentCategory
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: This should come from https://github.com/frequenz-floss/frequenz-client-common-python/. We still need to move/recreate the wrappers in that repo (they are not available yet).

@frequenz-floss/api-reporting-team @frequenz-floss/api-microgrid-team FYI as this is common to at least all these APIs

File: src/frequenz/client/dispatch/_client.py [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+from google.protobuf.timestamp_pb2 import Timestamp
+
+# from ._types import ForecastFeature, Forecasts, Location
+
+
+@dataclass
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]:
```suggestion
@dataclass(frozen=True, kw_only=True)
```

File: src/frequenz/client/dispatch/_client.py [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+            ts.FromDatetime(dt)
+            return ts
+
+        if interval_filter:
+            time_interval = dispatch_pb2.TimeIntervalFilter(
+                start_from=to_timestamp(interval_filter.start_from),
+                start_to=to_timestamp(interval_filter.start_to),
+                end_from=to_timestamp(interval_filter.end_from),
+                end_to=to_timestamp(interval_filter.end_to),
+            )
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: The idea is that your `TimeIntervalFilter` wrapper class has `from_proto()` and `to_proto()` methods. For now in the SDK we only have `from_proto()` ([example](https://github.com/frequenz-floss/frequenz-sdk-python/blob/91821a3e2719024c0b2a114d6a11aa692d3429d6/src/frequenz/sdk/microgrid/component/_component_data.py#L85-L112)) but I guess it could make sense to have the other too.

That said, to be honest, I don't like this approach, as the idea is to hide the underlying protocol we are using in the client, the idea is that the client user doesn't need to know about protobuf. Since conversion only needs to happen in the client internally, I would prefer an approach where we provide conversion functions separately from the wrapper class, similar to what marshmallow does, or even dataclass when converting to `dict` or `tuple`.

Maybe this is something to discuss in the next API meeting (or even SDK meeting, as this is a Python thing more than a API thing).

FYI @frequenz-floss/python-sdk-team

File: tests/test_dispatch_client.py [hunk 5]
```diff
@@ -2,4 +2,10 @@
 # Copyright © 2024 Frequenz Energy-as-a-Service GmbH
 
 """Tests for the frequenz.client.dispatch package."""
-import pytest
+
+
+def test_client() -> None:
+    """Test the dispatch client."""
```
Comment (llucax) [inline tests/test_dispatch_client.py]: That's a bold statement! :D

**Review by llucax — COMMENTED** [review]
Hi, some early feedback. About the SDK thing, if it is only a temporary hack to get things going, I'm good with it!

File: src/frequenz/client/dispatch/_client.py (hunk 4 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py]: I am not sure for this particular type it's worth having all that, given that it's only used in this function and no-where outside, and the conversion is extremely simple

Comment (llucax) [reply src/frequenz/client/dispatch/_client.py]: Right, maybe I picked the wrong example.

File: tests/test_dispatch_client.py (hunk 5 above)
Comment (Marenz) [reply tests/test_dispatch_client.py]: give it time :P

File: .github/workflows/ci.yaml
Comment (llucax) [inline .github/workflows/ci.yaml:1]: This should be split to a different commit.

File: src/frequenz/client/dispatch/_types.py [hunk 6]
```diff
@@ -0,0 +1,348 @@
...
+)
+
+ComponentSelector = list[int] | ComponentCategory
+"""A component selector specifying which components a dispatch targets.
+
+A component selector can be a list of component IDs or a component category.
+"""
+
+
+def _component_selector_from_protobuf(
```
Comment (llucax) [inline src/frequenz/client/dispatch/_types.py]: I would make this and the `to_protobuf` functions public (not starting with `_`) as you are importing it in another module. This whole module is private so it's not like the end user will have (direct) access to it.

File: src/frequenz/client/dispatch/_types.py [hunk 7]
```diff
@@ -0,0 +1,348 @@
...
+        pb_dispatch.microgrid_id = self.microgrid_id
+        pb_dispatch.type = self.type
+        pb_dispatch.create_time.FromDatetime(self.create_time)
+        pb_dispatch.update_time.FromDatetime(self.update_time)
+        pb_dispatch.start_time.FromDatetime(self.start_time)
+        pb_dispatch.duration = int(self.duration.total_seconds())
+        pb_dispatch.selector.CopyFrom(_component_selector_to_protobuf(self.selector))
+        pb_dispatch.is_active = self.is_active
+        pb_dispatch.is_dry_run = self.is_dry_run
+        # pb_dispatch.payload = self.payload
```
Comment (llucax) [inline src/frequenz/client/dispatch/_types.py]: Why is this commented out?

File: tests/test_dispatch_client.py [hunk 8]
```diff
@@ -2,4 +2,97 @@
...
+    dispatch_servicer = DispatchServicer()
+    dispatch_pb2_grpc.add_MicrogridDispatchServiceServicer_to_server(
+        dispatch_servicer, server
+    )
+    port = server.add_insecure_port("localhost:0")
+    await server.start()
+    try:
+        yield server, f"localhost:{port}"
+    finally:
+        await server.stop(None)
```
Comment (llucax) [inline tests/test_dispatch_client.py]: I should **not** use this mock service if it is really listening to a port, we did that in the SDK and it had a lot of issues, among them that multiple tests can be run at the same time. We should mock the server instead at the python level, patching the functions we are calling in our code.

All you need is to mock/patch (in the pyhon sense) this function used by the client:  `self._stub.ListMicrogridDispatches(request)` to return whatever mock data you want to have. We don't need to test the whole gRPC stuff, only our client code.

We could also do a test of the whole thing, but that should be an integration test, not a unit test.

File: tests/test_dispatch_client.py [hunk 9]
```diff
@@ -2,4 +2,97 @@
...
+
+
+@pytest.mark.asyncio
```
Comment (llucax) [inline tests/test_dispatch_client.py]: This mark shouldn't be necessary, usually pytest detects it automatically. Any reason to include it?

File: tests/test_dispatch_types.py [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+
+    for selector in (
+        [1, 2, 3],
+        [10, 20, 30],
+        ComponentCategory.BATTERY,
+        ComponentCategory.GRID,
+        ComponentCategory.METER,
+    ):
+        protobuf = _component_selector_to_protobuf(selector)  # type: ignore
+        assert _component_selector_from_protobuf(protobuf) == selector
```
Comment (llucax) [inline tests/test_dispatch_types.py]: Any reason not to use `pytest.mark.parametrize` for this? It will move some clutter out of the test itself.

Maybe you can borrow some ideas from here if you have more complex tests and you want, for example, to assign names to them, etc.:

https://github.com/frequenz-floss/pymdownx-superfence-filter-lines-python/blob/231d34c4f80ad00785b231d49e28453b5ccfb08e/tests/test_do_validate.py#L33-L127

File: tests/test_dispatch_types.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+def test_component_selector() -> None:
+    """Test the component selector."""
+    # pylint: disable=import-outside-toplevel
+    from frequenz.client.common.microgrid.components.components import ComponentCategory
+
+    from frequenz.client.dispatch._types import (
+        ComponentSelector,
+        _component_selector_from_protobuf,
+        _component_selector_to_protobuf,
+    )
```
Comment (llucax) [inline tests/test_dispatch_types.py]: Any reason for the imports inside the test?

File: tests/test_dispatch_types.py [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+
+    assert ComponentSelector is not None
```
Comment (llucax) [inline tests/test_dispatch_types.py]: Why this assert? How could it be `None`? I don't get this.

I guess maybe is a copy&paste thing from the very silly tests in the generated files in the API? In there it makes *some* sense because we are "testing" generated code, and generated code that is super dynamic, so maybe it could happen that you import something and it is `None`, but not even sure if they make sense there either.

**Review by llucax — COMMENTED** [review]
Mostly minor things, except for the gRPC mock server which caused us a lot of headaches in the SDK.

File: src/frequenz/client/dispatch/_types.py (hunk 7 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_types.py]: It caused trouble because the attribute is read-only. Haven'- investigated more

File: tests/test_dispatch_client.py (hunk 9 above)
Comment (Marenz) [reply tests/test_dispatch_client.py]: It was ChatGPTed :angel:

File: tests/test_dispatch_types.py (hunk 10 above)
Comment (Marenz) [reply tests/test_dispatch_types.py]: Tbh, I didn't feel the need to do more complex testing with these methods, they seem simple enough

File: tests/test_dispatch_types.py (hunk 11 above)
Comment (Marenz) [reply tests/test_dispatch_types.py]: I guess keeping things near where they are used. No deeper reason

File: src/frequenz/client/dispatch/_types.py (hunk 7 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_types.py]: Oh, weird. Maybe add a comment or create an issue about it?

File: tests/test_dispatch_types.py (hunk 11 above)
Comment (llucax) [reply tests/test_dispatch_types.py]: If there is no good reason, I would put it on the top-level, which is the most popular style and doesn't require an exception for `pylint`.

File: tests/test_dispatch_types.py (hunk 10 above)
Comment (llucax) [reply tests/test_dispatch_types.py]: Sure, you don't have to make it more complicated, but using `parametrize()` is actually simpler, you don't need to create classes for test cases, that was just a FYI in case it was useful. Just using parametrize is as simple as:

```py
from frequenz.client.common.microgrid.components.components import ComponentCategory

from frequenz.client.dispatch._types import (
    ComponentSelector,
    _component_selector_from_protobuf,
    _component_selector_to_protobuf,
)

@pytest.mark.parametrize("selector", [
        [1, 2, 3],
        [10, 20, 30],
        ComponentCategory.BATTERY,
        ComponentCategory.GRID,
        ComponentCategory.METER,
    ])
def test_component_selector(selector: list[int] | ComponentCategory) -> None:
    """Test the component selector."""
    protobuf = _component_selector_to_protobuf(selector)  # type: ignore
    assert _component_selector_from_protobuf(protobuf) == selector
```

Another advantage of using parametrize is it creates one test per parameter, so you get exactly which parameter failed and you can execute the test for one particular parameter only, also without it, the test will be aborted after the first failure, while using parametrize will continue testing other parameters (unless the fail early option is used).

If you think updating these tests here now are not worth it, I'm OKish with that, but I strongly suggest using `parametrize()` in future tests.

**Comment by Marenz** [comment]
@llucax this is almost ready, some mypy/pylint/flake8 complains, but everything else is working already. Feel free to review already.

File: pyproject.toml [hunk 13]
```diff
@@ -26,9 +26,17 @@ classifiers = [
...
+  "typing-extensions >= 4.6.1, < 5",
+# Use direct commit until release
+  "frequenz-api-dispatch @ git+https://github.com/frequenz-floss/frequenz-api-dispatch.git@223315c",
+#  "frequenz-api-dispatch >= 0.13.0, < 0.14",
+# Directly use unreleased commit until the next release
+  "frequenz-client-base @ git+https://github.com/frequenz-floss/frequenz-client-base-python.git@57d3093",
+#  "frequenz-client-base >= 0.1.0, < 0.2",
+# Directly use unreleased/unmerged commit until the first release
+  "frequenz-client-common @ git+https://github.com/frequenz-floss/frequenz-client-common-python.git@5e6e7b7",
+#  "frequenz-client-common == 0.1.0",
```
Comment (llucax) [inline pyproject.toml]: This is 100% the OCD in me talking, but any reason why the comments are not indented the same as the other lines? :grimacing:

File: src/frequenz/client/dispatch/_client.py [hunk 14]
```diff
@@ -0,0 +1,234 @@
...
+
+    # pylint: disable=too-many-arguments, too-many-locals
+    async def list(
+        self,
+        microgrid_id: int,
+        component_selectors: list[ComponentSelector] | None = None,
+        start_from: datetime | None = None,
+        start_to: datetime | None = None,
+        end_from: datetime | None = None,
+        end_to: datetime | None = None,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:56]: Not necessarily for this PR, but maybe in the future we want to add a `DatetimeRange` dataclass to pass pairs `from`/`to`.

File: src/frequenz/client/dispatch/_client.py [hunk 15]
```diff
@@ -0,0 +1,234 @@
...
+        is_active: bool | None = None,
+        is_dry_run: bool | None = None,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: Minor: I think attributes are more readable as `is_xxx`, but arguments read better using just `xxx` (so just `active` and `dry_run` in this case). I've taken [this recommendation from dart's style guide](https://dart.dev/effective-dart/design#prefer-a-non-imperative-verb-phrase-for-a-boolean-property-or-variable) and I find them very good and the resulting code very readable.

File: src/frequenz/client/dispatch/_client.py [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+        """
+        time_interval = None
+
+        def to_timestamp(dt: datetime | None) -> Timestamp | None:
+            if dt is None:
+                return None
+
+            ts = Timestamp()
+            ts.FromDatetime(dt)
+            return ts
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:95]: This might be useful to include in https://github.com/frequenz-floss/frequenz-client-base-python/, as all repos dealing with `datetime` will have to do it.

File: src/frequenz/client/dispatch/_client.py [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+
+        if start_from or start_to or end_from or end_to:
+            time_interval = PBTimeIntervalFilter(
+                start_from=to_timestamp(start_from),
+                start_to=to_timestamp(start_to),
+                end_from=to_timestamp(end_from),
+                end_to=to_timestamp(end_to),
+            )
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:103]: Just out of curiosity, this is some sort of optimization to send a message as small as possible over the wire, right? But it should be also possible to always create the PB object with all `None`s and it will have the same effect?

If so maybe you can add a comment to clarify it, so nobody thinks it is redundant in the future and removes the `if`.

File: src/frequenz/client/dispatch/_client.py [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+
+        selectors = []
+
+        for selector in component_selectors if component_selectors else []:
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: You can shorten this a bit with:

```suggestion
        for selector in component_selectors or []:
```

But it is probably even simpler to make the argument a tuple and have a `()` default instead.

File: src/frequenz/client/dispatch/_client.py [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+            svc_addr: Address of the service to connect to.
+        """
+        self._svc_addr = svc_addr
+        self._stub = dispatch_pb2_grpc.MicrogridDispatchServiceStub(grpc_channel)
+
+    # pylint: disable=too-many-arguments, too-many-locals
+    async def list(
+        self,
+        microgrid_id: int,
+        component_selectors: list[ComponentSelector] | None = None,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]:
```suggestion
        component_selectors: Iterator[ComponentSelector] = (),
```

File: src/frequenz/client/dispatch/_client.py [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+        start_from: datetime | None = None,
+        start_to: datetime | None = None,
+        end_from: datetime | None = None,
+        end_to: datetime | None = None,
+        is_active: bool | None = None,
+        is_dry_run: bool | None = None,
+    ) -> list[Dispatch]:
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: I think this should return an async iterator instead.
```suggestion
    ) -> AsyncIterator[Dispatch]:
```

So users can just do:

```py
async for dispatch in client.list():
    ...
```

Once the return type is `AsyncIterator`, you can just `yield` each value in the function body. This will be particularly important when we add pagination, so users don't need to wait (and hold in memory) all the pages before starting to process dispatches.

File: src/frequenz/client/dispatch/_client.py [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+            selectors=selectors,
+            time_interval=time_interval,
+            is_active=is_active,
+            is_dry_run=is_dry_run,
+        )
+        request = DispatchListRequest(microgrid_id=microgrid_id, filter=filters)
+
+        response: Awaitable[DispatchList]
+        response = self._stub.ListMicrogridDispatches(request)  # type: ignore
+        return list(map(Dispatch.from_protobuf, (await response).dispatches))
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]:
```suggestion
        response = await self._stub.ListMicrogridDispatches(request)
        for dispatch in response.dispatches:
            yield Dispatch.from_protobuf(dispatch)
```

File: src/frequenz/client/dispatch/_client.py [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+        self,
+        microgrid_id: int,
+        _type: str,
+        start_time: datetime,
+        duration: timedelta,
+        selector: ComponentSelector,
+        is_active: bool,
+        is_dry_run: bool,
+        payload: dict[str, Any],
+        recurrence: RecurrenceRule,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]:
```suggestion
        recurrence: RecurrenceRule | None = None,
```
?

In this case I'm pretty sure recurrence rules are optional, but we should also document it in the protobuf spec.

File: src/frequenz/client/dispatch/_client.py [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+        response = self._stub.ListMicrogridDispatches(request)  # type: ignore
+        return list(map(Dispatch.from_protobuf, (await response).dispatches))
+
+    async def create(
+        self,
+        microgrid_id: int,
+        _type: str,
+        start_time: datetime,
+        duration: timedelta,
+        selector: ComponentSelector,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:128]:
```suggestion
        selector: ComponentSelector | None = None,
```
? (and move it to the end)

Every sub-message in protobuf is implicitly `optional`. Unless the server will check its existence and fail if it is not provided. I checked the docs and it doesn't seem to say what happens if there is no selector, but I would assume all dispatches will be returned.

@stefan-brus-frequenz can you confirm this? We should also add it to the proto documentation.

File: src/frequenz/client/dispatch/_client.py [hunk 24]
```diff
@@ -0,0 +1,234 @@
...
+            selector: The component selector for the dispatch.
+            is_active: The is_active status of the dispatch.
+            is_dry_run: The is_dry_run status of the dispatch.
+            payload: The payload of the dispatch.
+            recurrence: The recurrence rule of the dispatch.
+        """
+        request = DispatchCreateRequest(
+            microgrid_id=microgrid_id,
+            type=_type,
+            start_time=start_time,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:156]: From https://github.com/frequenz-floss/frequenz-api-dispatch/blob/223315c6038778acf0b2384f881c69155d96a49c/proto/frequenz/api/dispatch/v1/dispatch.proto#L324C1-L328C44, we should check for this:

```proto
  // The start time
  // When creating a dispatch, ensure that the starting timestamp is set to
  // the current time or any future time.
  // Timestamps earlier than the current time are not allowed.
  google.protobuf.Timestamp start_time = 3;
```

File: src/frequenz/client/dispatch/_client.py [hunk 25]
```diff
@@ -0,0 +1,234 @@
...
+    async def create(
+        self,
+        microgrid_id: int,
+        _type: str,
+        start_time: datetime,
+        duration: timedelta,
+        selector: ComponentSelector,
+        is_active: bool,
+        is_dry_run: bool,
+        payload: dict[str, Any],
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]:
```suggestion
        payload: dict[str, Any] | None = None,
```
? Also pretty sure the payload is optional.

File: src/frequenz/client/dispatch/_client.py [hunk 26]
```diff
@@ -0,0 +1,234 @@
...
+
+    async def create(
+        self,
+        microgrid_id: int,
+        _type: str,
+        start_time: datetime,
+        duration: timedelta,
+        selector: ComponentSelector,
+        is_active: bool,
+        is_dry_run: bool,
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py]: I'd probably also make these optional, defaulting to `True` and `False` respectively.

@stefan-brus-frequenz any opinion on this? Since these are not `optional` in protobuf, if omitted they will default to something, which I would guess is `False`?

File: src/frequenz/client/dispatch/_client.py [hunk 27]
```diff
@@ -0,0 +1,234 @@
...
+            payload=payload,
+            recurrence=recurrence,
+        ).to_protobuf()
+
+        await self._stub.CreateMicrogridDispatch(request)  # type: ignore
+
+    async def update(
+        self,
+        dispatch_id: int,
+        new_fields: dict[str, Any],
```
Comment (llucax) [inline src/frequenz/client/dispatch/_client.py:170]:
```suggestion
        **new_fields: Any,
```

But I would actually list explicitly everything that can be updated here, otherwise we lose type safety and also we allow introducing bugs because of typos (`update(start_tim=...)` won't be caught by `mypy`) and the code is quite convoluted with the huge `match` statement. Actually now that I see it the `**new_fields: Any` won't work because you use `.` in keys.

For this PR I'm fine with this but I think we should definitely do it more static and type checked in the future.

File: src/frequenz/client/dispatch/_types.py [hunk 28]
```diff
@@ -0,0 +1,442 @@
...
+        pb_request.is_active = self.is_active
+        pb_request.is_dry_run = self.is_dry_run
+        pb_request.payload.update(self.payload)
+        pb_request.recurrence.CopyFrom(self.recurrence.to_protobuf())
+
+        return pb_request
+
+
+@dataclass(kw_only=True)
+class Dispatch(DispatchCreateRequest):
```
Comment (llucax) [inline src/frequenz/client/dispatch/_types.py]: I wouldn't do this, besides both could diverge, this means someone could accidentally pass a `Dispatch` object where a `DispatchCreateRequest` is expected. I know it should be unlikely because `DispatchCreateRequest` should only be used internally, but that's actually another issue, we should better separate classes that are supposed to be public from classes that should be private, and public classes shouldn't inherit from private classes as users shouldn't have access to the private class.

File: src/frequenz/client/dispatch/_types.py [hunk 29]
```diff
@@ -0,0 +1,442 @@
...
+
+        Returns:
+            The initialized dispatch.
+        """
+        return Dispatch(
+            id=_id,
+            create_time=create_time,
+            update_time=update_time,
+            **_request.__dict__,
+        )
```
Comment (llucax) [inline src/frequenz/client/dispatch/_types.py]: This is not even used, right?

**Review by llucax — COMMENTED** [review]
Besides the minor (optional) comments, I'm mostly OK to merge this as is, but there are a few more *major* points that I think we should change in the future:

* Make `list` async iterable
* Make `update` type safe
* Separate public from private classes more clearly (I would put them in different files)
* Make sure public classes don't inherit from private classes

PS: I didn't had a look at the tests yet

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:170]: >  the code is quite convoluted with the huge match statement

It would look similar with any other solution as I have to take care / convert every (type of ) member individually.. if it's not a `match` statement it would be a `if para: ...` instead for each parameter..

File: src/frequenz/client/dispatch/_types.py (hunk 29 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_types.py]: for tests mostly

File: src/frequenz/client/dispatch/_types.py (hunk 28 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_types.py]: I really hate the repetition that that will cause...

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:170]: Yeah, is true that you will still check which ones are set to be updated and which are not one by one, but you could at least group `RecurrenceRuleUpdate` in wrapper, so that can be omitted as a group if nothing has to be changed for example, but the main point here for me is still type safety and being able to get errors at *`mypy`time* :D

Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:170]: I though about other options, using `TypedDict` for example, to avoid the bloat, but all have their issues, I guess this is something we might need to think more about, so I would leave it as in this PR and create an issue about it. I think how we do updates should also be consistent among all API clients, so is something we might want to discuss with the other API teams.

File: src/frequenz/client/dispatch/_types.py (hunk 29 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_types.py]: Why not adding it to the `tests/` folder then? I you need to reuse it from different files you can just add something like `tests/utils.py`.

File: src/frequenz/client/dispatch/_client.py (hunk 19 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py]: `ComponentSelector` itself is `ComponentSelector = list[int] | ComponentCategory` should that also become an Iterator?

File: src/frequenz/client/dispatch/_client.py (hunk 17 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:103]: > But it should be also possible to always create the PB object with all Nones and it will have the same effect?

I'd need to research that to be sure, right now, no idea

File: src/frequenz/client/dispatch/_client.py (hunk 23 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:128]: This is the create function, it doesn't return anything :)

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:170]: A data class with every type optional as a parameter could be a way.

File: src/frequenz/client/dispatch/_types.py (hunk 28 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_types.py]: Again, this is supposed to be stable, is a one time thing.

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_client.py:170]: Yeah, we would need to use a sentinel to indicate if a value was set or not (because you might need to update something from a value to `None`, so `None` can't mean "this attribute is not updated") but that still requires you to check one by one every attribute to see if it was overridden or not. `dict`s are the only ones that have the nice property of encoding the absence of a field natively, so you can just iterate over the existing keys to set the attributes in the message.

File: src/frequenz/client/dispatch/_client.py (hunk 19 above)
Comment (llucax) [reply src/frequenz/client/dispatch/_client.py]: you mean the `list[int]`? Nah, I think that should still be a list, because otherwise it would be more error prone when using as an attribute, if people iterate over the attribute without realizing they "consumed" the iterator.

In this case is safe to take an iterator as argument because we know we only need to iterate it and once. The downside of making it an iterator is if at some point we need to do more we either need to make a list of the iterator ourselves, or change the interface to take a list. But I guess this is unlikely to happen.

File: src/frequenz/client/dispatch/_client.py (hunk 14 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:56]: https://github.com/frequenz-floss/frequenz-client-dispatch-python/issues/15

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:170]: I could also use the top suggestion and use `_` instead of `.` as path separator..

Comment (leandro-lucarella-frequenz) [reply src/frequenz/client/dispatch/_client.py:170]: For what exactly?

File: src/frequenz/client/dispatch/_client.py (hunk 23 above)
Comment (leandro-lucarella-frequenz) [reply src/frequenz/client/dispatch/_client.py:128]: I would still like to confirm users can't create a new dispatch without a `ComponentSelector`. @stefan-brus-frequenz

**Review by leandro-lucarella-frequenz — APPROVED** [review]

File: src/frequenz/client/dispatch/_client.py (hunk 27 above)
Comment (Marenz) [reply src/frequenz/client/dispatch/_client.py:170]: For making `**new_fields: Any` possible
